examples/main.tf

```diff
@@ -0,0 +1,128 @@
+*
+* This example also shows how to create:
+* - several users with a single module block,
+* - a group and add group members to it,
+* - a policy pertaining to a compartment and group,
+* - some more directives to show dynamic groups and policy for it.
+*
+* Note: The compartment resource internally resolves name collisions and returns a reference to the preexisting compartment.
+* All resources created by this example can be deleted by using the Terraform destroy command.
+ */
+
+module "iam_compartment" {
+
+  source                  = "oracle-terraform-modules/iam/oci//modules/iam-compartment"
+  version                 = "2.1.1"
+  tenancy_ocid            = var.tenancy_ocid
+  compartment_id          = var.tenancy_ocid # define the parent compartment. Creation at tenancy root if omitted
+  compartment_name        = "tf_example_compartment"
+  compartment_description = "compartment at root level created - terraformed"
+  compartment_create      = true # if false, a data source with a matching name is created instead
+  enable_delete           = true # if false, on `terraform destroy`, compartment is deleted from the terraform state but not from oci 
+}
+
+module "iam_subcompartment1" {
+  source                  = "oracle-terraform-modules/iam/oci//modules/iam-compartment"
+  version                 = "2.1.1"
+  tenancy_ocid            = var.tenancy_ocid
+  compartment_id          = module.iam_compartment.compartment_id # define the parent compartment. Here we make reference to the previous module
+  compartment_name        = "tf_example_subcompartment1"
+  compartment_description = "subcompartment created below tf_example_compartment - terraformed"
+  compartment_create      = true # if false, a data source with a matching name is created instead
+  enable_delete           = true # if false, on `terraform destroy`, compartment is deleted from the terraform state but not from oci 
+}
+
+module "iam_subcompartment2" {
+  source                  = "oracle-terraform-modules/iam/oci//modules/iam-compartment"
+  version                 = "2.1.1"
+  tenancy_ocid            = var.tenancy_ocid
+  compartment_id          = module.iam_compartment.compartment_id # define the parent compartment. Here we make reference to the previous module
+  compartment_name        = "tf_example_subcompartment2"
+  compartment_description = "subcompartment created below tf_example_compartment - terraformed"
+  compartment_create      = true # if false, a data source with a matching name is created instead
+  enable_delete           = true # if false, on `terraform destroy`, compartment is deleted from the terraform state but not from oci 
+}
+
+module "iam_users" {
+  source       = "oracle-terraform-modules/iam/oci//modules/iam-user"
+  version      = "2.1.1"
+  tenancy_ocid = var.tenancy_ocid
+  users = [
+    { # user1
+      name        = "tf_example_user1@example.com"
+      description = "user1 created by terraform"
+      email       = null
```

Comment:
why it is null here?

Reply:
This is to illustrate how to provision a user without providing an email, in contrast with the two other users declared in this example module configuration.

- "email" is an optional field for oci_identity_user in the oci provider, and we want to keep the possibility to provision a user without providing an email.
- But the "email" key needs to be present in the module definition block.

Providing a `null` value is interpreted internally by the module to an empty string (see module/iam-user/main.tf line 20).

`email          = var.users[count.index].email == null ? "" : var.users[count.index].email`